File: examples/src/rpi_i2c_6.py
import i2c
import ad7991
import max518
import time

# Instantiate the ADC and DAC device
adc = ad7991.device( i2c.device("/dev/i2c-1"), 0)
dac = max518.device( i2c.device("/dev/i2c-1"), 0 )

# Set the number of ADC channels
adc.set_channel(ad7991.TWO_CHAN)

i = 0;

while True:

    # Set the DAC levels in opposite directions
    dac.set_level( 0, i )
    dac.set_level( 1, 255-i )
    
    # Allow time for the loop
    time.sleep(1)
    
    # Get the value for channel 0
    channel, value = adc.get_level()
    print ('Channel = ' + channel + ' ADC value = ' + value + ' DAC value = ' + i)

    # Get the value for channel 1
    channel, value = adc.get_level()
    print ('Channel = ' + channel + ' ADC value = ' + value + ' DAC value = ' + str(255-i))
    
    # Increment DAC value
    i = i + 1

# Destroy the device
del adc

File: examples/src/rpi_i2c_6.c
#include <scriptml/ad7991.h>
#include <scriptml/max518.h> 

#include <scriptml/i2c.h>
#include <stdint.h>
#include <test_print.h>
#include <stdio.h>

int main()
{	
    /* Instantiate the ADC and the DAC on the I2C bus */
    sml_ad7991_t adc = sml_ad7991_create( sml_i2c_open( "/dev/i2c-1" ), 0 );
	sml_max518_t dac = sml_max518_create( sml_i2c_open( "/dev/i2c-1" ), 0 );
    
	uint8_t rd_buf[2], wr_buf[1];
    
    /* Get samples from Vin0 and Vin1 alternately, all other configs left at defaults */
	wr_buf[0] = 0x30;

    /* Write the config value to the ADC */
	sml_ad7991_write( adc, wr_buf, 0, 1 );

	sleep(1);
    
    int i = 0;
    
    /* Define two registers to hold the 16 bit ADC register contents. The 12 LSB are
       the ADC value, and the two bits above that, the channel identity */
    uint8_t MSB, LSB;
    uint8_t ChannelNumber;
    
    int adcValue;
    
    /* Set the DAC vale and then read contents of ADC */
    while(1){
        
        /* Set the DAC level for channels 1 and 2 */
        sml_max518_set_level( dac, 0, i );
        sml_max518_set_level( dac, 1, 255 - i );  
        
 
        
        /* Pause to allow DAC to settle (much longer than needed here to also slow display) */
        sleep(1);
        
        /* Read from the ADC, two bytes */
        sml_ad7991_read( adc, rd_buf, 0, 2 );
        MSB = rd_buf[0];
        LSB = rd_buf[1];
        
        /* Mask and bit shift to get the channel identifier */
        ChannelNumber = (MSB & 0x30) >> 4;
        
        /* Get the 12 bit ADC value */
        adcValue = LSB + ((MSB & 0x0F) << 8);
        
        /* Print the values */
        printf("Channel = %d, Value = %d, DAC0 value = %d, DAC1 value = %d\n", ChannelNumber, adcValue, i, 255-i);
        
        /* increment the DAC set value for the next iteration */
        i++;
    }
	return 0;
}

